File: src/tac.ml
type temp = int
type label = int
type string_id = int

type operand =
  | Void
  | Temp of temp
  | Proc of Symbol.t
  | Param of Symbol.t
  | I32 of int
  | String of string_id

type binop =
  | Add
  | Sub
  | Mul
  | Div

type instruction =
  | Binop of
      { dst : temp
      ; ty : Type.t
      ; op : binop
      ; lhs : operand
      ; rhs : operand
      }
  | Call of
      { dst : temp option
      ; ty : Type.t
      ; callee : operand
      ; args : operand list
      }

type terminator =
  | Ret of operand option
  | Jmp of label
  | Branch of
      { cond : operand
      ; if_true : label
      ; if_false : label
      }

type block =
  { label : label
  ; instrs : instruction list
  ; terminator : terminator
  }

type linkage =
  | Internal
  | Export
  | Extern
  | Extern_Export

type proc =
  { sym : Symbol.t
  ; linkage : linkage
  ; params : Symbol.t list
  ; return_type : Type.t
  ; entry : label option
  ; blocks : block list
  }

type string_literal =
  { id : string_id
  ; value : string
  }

type program =
  { strings : string_literal list
  ; procs : proc list
  }

let binop_to_string = function
  | Add -> "add"
  | Sub -> "sub"
  | Mul -> "mul"
  | Div -> "Div"

let binop_of_token = function
  | Token.Plus          -> Add
  | Token.Minus         -> Sub
  | Token.Asterisk      -> Mul
  | Token.Forward_Slash -> Div
  | _ -> failwith "unsupported TAC binary operator"

let operand_to_string = function
  | Void      -> "void"
  | Temp id   -> "%" ^ string_of_int id
  | Proc p   -> "@proc" ^ string_of_int p.id
  | Param p   -> "%arg" ^ string_of_int p.id
  | I32 v     -> string_of_int v
  | String id -> "@str" ^ string_of_int id

let instruction_to_string = function
  | Binop {dst; ty; op; lhs; rhs} ->
     Printf.sprintf "%%%d = %s %s %s, %s"
       dst
       (binop_to_string op)
       (Type.to_string ty)
       (operand_to_string lhs)
       (operand_to_string rhs)
  | Call {dst; ty; callee; args} ->
     let result = match dst with
       | Some dst -> Printf.sprintf "%%%d = " dst
       | None -> ""
     in
     Printf.sprintf "%scall %s %s(%s)"
       result
       (Type.to_string ty)
       (operand_to_string callee)
       (args |> List.map operand_to_string |> String.concat ", ")

let terminator_to_string = function
  | Ret None -> "ret"
  | Ret (Some v) ->
     "ret " ^ operand_to_string v
  | Jmp lbl ->
     "jmp L" ^ string_of_int lbl
  | Branch {cond; if_true; if_false} ->
     Printf.sprintf "branch %s, L%d, L%d"
       (operand_to_string cond)
       if_true
       if_false

let block_to_string blk =
  let instructions =
    List.map (fun i -> "  " ^ instruction_to_string i) blk.instrs
  in
  String.concat "\n"
    (("L" ^ string_of_int blk.label ^ ":") ::
       instructions @
       ["  " ^ terminator_to_string blk.terminator])

let linkage_to_string = function
  | Internal -> ""
  | Export -> "export "
  | Extern -> "extern "
  | Extern_Export -> "extern export "

let is_extern = function
  | Extern | Extern_Export -> true
  | Internal | Export -> false

let proc_to_string proc =
  let params =
    proc.params
    |> List.map (fun (sym : Symbol.t) ->
           "%arg" ^ string_of_int sym.id ^ ": " ^ Type.to_string sym.ty)
    |> String.concat ", "
  in
  if is_extern proc.linkage then
    Printf.sprintf "%sproc %s(%s%s): %s;"
      (linkage_to_string proc.linkage)
      proc.sym.name
      params
      (match proc.sym.ty with Proc {variadic = true; _} -> " ... " | _ -> "")
      (Type.to_string proc.return_type)
  else
    let blks =
      proc.blocks
      |> List.map block_to_string
      |> String.concat "\n"
    in
    Printf.sprintf "%sproc %s(%s%s): %s {\n%s\n}"
      (linkage_to_string proc.linkage)
      proc.sym.name
      params
      (match proc.sym.ty with Proc {variadic = true; _} -> ", ..." | _ -> "")
      (Type.to_string proc.return_type)
      blks

let program_to_string prog =
  let strings = List.map (
                    fun s -> Printf.sprintf "@str%d = \"%s\""
                               s.id (String.escaped s.value)
                  ) prog.strings
  in
  let procs = prog.procs |> List.map proc_to_string in
  String.concat "\n\n" (strings @ procs)

File: src/tac.cpp
#include "tac.hpp"

#include <exception>
#include <format>
#include <utility>

struct TAC_error : public std::exception {
        std::string msg;

        TAC_error(std::string msg)
                : msg(std::move(msg)) {}

        const char *
        what(void) const noexcept override
        {
                return this->msg.c_str();
        }
};

struct unknown_binary_operator_error : TAC_error {
        token_kind k;

        unknown_binary_operator_error(token_kind k)
                : TAC_error(std::format("cannot derive TAC binary operator from token `%s'",
                                        token_kind_to_cstring(k))),
                k(k) {}
};

static std::string
TAC_binary_operator_to_string(TAC_binary_operator op)
{
        switch (op) {
        case TAC_BINARY_OPERATOR_ADD:           return "add";
        case TAC_BINARY_OPERATOR_SUB:           return "sub";
        case TAC_BINARY_OPERATOR_MUL:           return "mul";
        case TAC_BINARY_OPERATOR_DIV:           return "div";
        case TAC_BINARY_OPERATOR_MOD:           return "mod";
        case TAC_BINARY_OPERATOR_OR:            return "or";
        case TAC_BINARY_OPERATOR_AND:           return "and";
        case TAC_BINARY_OPERATOR_XOR:           return "xor";
        case TAC_BINARY_OPERATOR_LESS:          return "lt";
        case TAC_BINARY_OPERATOR_GREATER:       return "gt";
        case TAC_BINARY_OPERATOR_LESS_EQUAL:    return "le";
        case TAC_BINARY_OPERATOR_GREATER_EQUAL: return "ge";
        case TAC_BINARY_OPERATOR_EQUAL:         return "eq";
        case TAC_BINARY_OPERATOR_NOT_EQUAL:     return "ne";
        }
        std::unreachable();
}

static TAC_binary_operator
token_kind_to_TAC_binary_operator(token_kind k)
{
        switch (k) {
        case TOKEN_KIND_PLUS:               return TAC_BINARY_OPERATOR_ADD;
        case TOKEN_KIND_HYPHEN:             return TAC_BINARY_OPERATOR_SUB;
        case TOKEN_KIND_ASTERISK:           return TAC_BINARY_OPERATOR_MUL;
        case TOKEN_KIND_FORWARD_SLASH:      return TAC_BINARY_OPERATOR_DIV;
        case TOKEN_KIND_PERCENT:            return TAC_BINARY_OPERATOR_MOD;
        case TOKEN_KIND_PIPE:               return TAC_BINARY_OPERATOR_OR;
        case TOKEN_KIND_AMPERSAND:          return TAC_BINARY_OPERATOR_AND;
        case TOKEN_KIND_UPTICK:             return TAC_BINARY_OPERATOR_XOR;
        case TOKEN_KIND_LESSTHAN:           return TAC_BINARY_OPERATOR_LESS;
        case TOKEN_KIND_GREATERTHAN:        return TAC_BINARY_OPERATOR_GREATER;
        case TOKEN_KIND_LESSTHAN_EQUALS:    return TAC_BINARY_OPERATOR_LESS_EQUAL;
        case TOKEN_KIND_GREATERTHAN_EQUALS: return TAC_BINARY_OPERATOR_GREATER_EQUAL;
        case TOKEN_KIND_EQUALS:             return TAC_BINARY_OPERATOR_EQUAL;
        case TOKEN_KIND_BANG_EQUALS:        return TAC_BINARY_OPERATOR_NOT_EQUAL;
        default:                            throw unknown_binary_operator_error(k);
        }
        std::unreachable();
}

static std::string
TAC_operand_to_string(const TAC_operand *const op)
{
        switch (op->k) {
        case TAC_OPERAND_KIND_VOID:   return "void";
        case TAC_OPERAND_KIND_TEMP:   return std::format("%%%d",      ((TAC_operand_temp *)op)->temp);
        case TAC_OPERAND_KIND_PROC:   return std::format("@proc%d",   ((TAC_operand_proc *)op)->sym->id);
        case TAC_OPERAND_KIND_PARAM:  return std::format("%%arg%d",   ((TAC_operand_param *)op)->sym->id);
        case TAC_OPERAND_KIND_LOCAL:  return std::format("%%local%d", ((TAC_operand_local *)op)->sym->id);
        case TAC_OPERAND_KIND_I32:    return std::format("%d",        ((TAC_operand_i32 *)op)->i);
        case TAC_OPERAND_KIND_STRING: return std::format("@str%d",    ((TAC_operand_string *)op)->id);
        }
        std::unreachable();
}

static std::string
TAC_instruction_to_string(const TAC_instruction *const ins)
{
        switch (ins->k) {
        case TAC_INSTRUCTION_KIND_BINOP: {
                const TAC_instruction_binop *const b = (TAC_instruction_binop *)ins;
                return std::format("%%%d = %s %s %s, %s",
                                   b->dst,
                                   TAC_binary_operator_to_string(b->op),
                                   type_to_string(b->ty),
                                   TAC_operand_to_string(b->lhs),
                                   TAC_operand_to_string(b->rhs));
        } break;
        case TAC_INSTRUCTION_KIND_LOAD: {
                const TAC_instruction_load *const l = (TAC_instruction_load *)ins;
                return std::format("%%%d = load %s %s",
                                   l->dst,
                                   type_to_string(l->ty),
                                   TAC_operand_to_string(l->src));
        } break;
        case TAC_INSTRUCTION_KIND_STORE: {
                const TAC_instruction_store *const s = (TAC_instruction_store *)ins;
                return std::format("store %s %s, %s",
                                   type_to_string(s->ty),
                                   TAC_operand_to_string(s->src),
                                   TAC_operand_to_string(s->dst));
        } break;
        }
        std::unreachable();
}

static std::string
TAC_terminator_to_string(const TAC_terminator *const t)
{
        switch (t->k) {
        case TAC_TERMINATOR_KIND_RET: {
                const TAC_terminator_ret *const r = (TAC_terminator_ret *)t;
                if (r->value == std::nullopt)
                        return "ret";
                return std::format("ret %s", TAC_operand_to_string(r->value.value()));
        } break;
        case TAC_TERMINATOR_KIND_JMP: {
                const TAC_terminator_jmp *const j = (TAC_terminator_jmp *)t;
                return std::format("jmp L%d", j->label);
        } break;
        case TAC_TERMINATOR_KIND_BRANCH: {
                const TAC_terminator_branch *const b = (TAC_terminator_branch *)t;
                return std::format("branch %s, L%d, L%d",
                                   TAC_operand_to_string(b->cond),
                                   b->if_true,
                                   b->if_false);
        } break;
        }
}

static std::string
TAC_block_to_string(const TAC_block *const blk)
{
        std::string res;

        res = std::format("L%d:\n", blk->label);

        for (size_t i = 0; i < blk->instructions.size(); ++i) {
                if (i != 0)
                        res += "\n";
                res += "  " + TAC_instruction_to_string(blk->instructions.at(i));
        }

        res += TAC_terminator_to_string(blk->terminator);

        return res;
}

static std::string
TAC_linkage_to_string(TAC_linkage l)
{
        switch (l) {
        case TAC_LINKAGE_INTERNAL:      return "";
        case TAC_LINKAGE_EXPORT:        return "export ";
        case TAC_LINKAGE_EXTERN:        return "extern ";
        case TAC_LINKAGE_EXTERN_EXPORT: return "export extern ";
        }
}

static inline
int is_extern(TAC_linkage l)
{
        return l >= TAC_LINKAGE_EXTERN;
}

static std::string
TAC_proc_to_string(const TAC_proc *const p)
{
        std::string linkage;
        std::string name;
        std::string params;
        std::string variadic;
        std::string return_type;
        std::string blocks;

        linkage = TAC_linkage_to_string(p->linkage);
        name    = p->sym->name;

        for (size_t i = 0; i < p->params.size(); ++i) {
                if (i != 0)
                        params += ", ";
                const symbol *const sym = p->params.at(i);
                params += std::format("%%arg%d: %s",
                                      sym->id, type_to_string(sym->ty));
        }

        variadic = ((type_proc *)p)->variadic ? "..." : "";

        return_type = type_to_string(p->return_type);

        for (size_t i = 0; i < p->blocks.size(); ++i) {
                if (i != 0)
                        blocks += "\n";
                blocks += TAC_block_to_string(p->blocks.at(i));
        }

        return linkage + "proc " + name + "(" + params + variadic + ": " + return_type + "{\n" + blocks + "}";
}

static std::string
TAC_string_value_to_string(const std::string &value)
{
        std::string result;

        for (size_t i = 0; i < value.size(); ++i) {
                const unsigned char c = value.at(i);
                switch (c) {
                case '\b': result += "\\b"; break;
                case '\t': result += "\\t"; break;
                case '\n': result += "\\n"; break;
                case '\r': result += "\\r"; break;
                case '"':  result += "\\\""; break;
                case '\\': result += "\\\\"; break;
                default:
                        if (c < 32 || c >= 127)
                                result += std::format("\\{:03d}", c);
                        else
                                result += c;
                        break;
                }
        }

        return result;
}

std::string
TAC_program_to_string(const TAC_program *const p)
{
        std::string result;
        bool first = true;

        for (size_t i = 0; i < p->strings.size(); ++i) {
                const TAC_string &s = p->strings.at(i);
                if (!first)
                        result += "\n\n";
                result += std::format("@str{} = \"{}\"",
                                      s.id, TAC_string_value_to_string(s.value));
                first = false;
        }

        for (size_t i = 0; i < p->procs.size(); ++i) {
                const TAC_proc &proc = p->procs.at(i);
                if (!first)
                        result += "\n\n";
                result += TAC_proc_to_string(&proc);
                first = false;
        }

        return result;
}
